File: leetcode/0073.SetMatrixZeroes/Solution.cpp
#include "../../include/global.hpp"
#include <bits/stdc++.h>

using namespace std;

class Solution {
  public:
    void setZeroes(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<int> mm(m, 1);
        vector<int> nn(n, 1);

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == 0) {
                    mm[i] = 0;
                    nn[j] = 0;
                }
            }
        }
        // cout << matrix << endl;

        for (int i = 0; i < m; i++) {
            if (mm[i] == 0) {
                for (int j = 0; j < n; j++) {
                    matrix[i][j] = 0;
                }
            }
        }
        // cout << matrix << endl;

        for (int j = 0; j < n; j++) {
            if (nn[j] == 0) {
                for (int i = 0; i < m; i++) {
                    matrix[i][j] = 0;
                }
            }
        }
    }
};

int main() {
    Solution s;
    // vector<vector<int>> matrix = {
    //     {1, 1, 1},
    //     {1, 0, 1},
    //     {1, 1, 1}};
    vector<vector<int>> matrix = {
        {0, 1, 2, 0},
        {3, 4, 5, 2},
        {1, 3, 1, 5}};
    s.setZeroes(matrix);
    cout << matrix << endl;
    return 0;
}

File: leetcode/0073.SetMatrixZeroes/Solution.go
package main

import "fmt"

func setZeroes(matrix [][]int) {
	firstRowHasZero := false
	firstColumnHasZero := false
	for n := 0; n < len(matrix[0]); n++ {
		if matrix[0][n] == 0 {
			firstRowHasZero = true
		}
	}
	for n := 0; n < len(matrix); n++ {
		if matrix[n][0] == 0 {
			firstColumnHasZero = true
		}
	}
	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[i]); j++ {
			if matrix[i][j] == 0 {
				matrix[i][0] = 0
				matrix[0][j] = 0
			}
		}
	}
	for i := 1; i < len(matrix); i++ {
		if matrix[i][0] == 0 {
			for j := 1; j < len(matrix[i]); j++ {
				matrix[i][j] = 0
			}
		}
	}
	for j := 1; j < len(matrix[0]); j++ {
		if matrix[0][j] == 0 {
			for i := 1; i < len(matrix); i++ {
				matrix[i][j] = 0
			}
		}
	}
	if firstRowHasZero == true {
		for n := 0; n < len(matrix[0]); n++ {
			matrix[0][n] = 0
		}
	}
	if firstColumnHasZero == true {
		for n := 0; n < len(matrix); n++ {
			matrix[n][0] = 0
		}
	}
}

func main() {
	// input := [][]int{{1, 1, 1}, {1, 0, 1}, {1, 1, 1}}
	input := [][]int{{0}, {1}}
	setZeroes(input)
	fmt.Println(input)
}
